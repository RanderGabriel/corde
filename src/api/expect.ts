import { expectMatchWithNot, MatchesWithNot } from "./matcher";

/**
 * Receives wich command will be tested.
 *
 * @param commandName Command name.
 *
 * @description Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @returns The **Compare** object, where will handle
 * the type of response is expected.
 */
export function expect(commandName: string): MatchesWithNot {
  return expectMatchWithNot(commandName);
}
