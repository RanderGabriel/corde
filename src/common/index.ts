export * from './cordeBot';
export * from './runtime';
export * from './testColletor';
