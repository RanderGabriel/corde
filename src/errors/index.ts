export * from './filesNotFoundError';
export * from './channelTypeNotSupportedError';
export * from './guildNotFoundError';
export * from './parameterNotFoundError';
export * from './configFileNotFoundError';
export * from './missingPropertyError';
export * from './runtimeError';
