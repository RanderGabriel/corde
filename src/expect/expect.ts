import { VoidLikeFunction } from "../types";
import { ExpectMatchesWithNot, MatchWithNot } from "./matcher";

/**
 * Invalid declaration of a test matcher.
 *
 * @since 3.0
 */
export function expect(): void;
/**
 * Receives which command will be tested.
 *
 * Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
 *
 * @returns An object with all possible tests to be done
 * in the bot.
 *
 * @since 1.0
 */
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable: T,
): MatchWithNot;
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable?: T,
): MatchWithNot | void {
  if (commandNameResolvable == null) {
    return;
  }
  return new ExpectMatchesWithNot(commandNameResolvable);
}
