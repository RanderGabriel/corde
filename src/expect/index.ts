import { RoleMatches, MessageMatches } from "./matcher";
import { Expect, MacherContructorArgs } from "./types";

function getMessageMatchers(): string[] {
  return getFunctions(MessageMatches);
}

function getRoleMatchers(): string[] {
  return getFunctions(RoleMatches);
}

function getFunctions<T>(type: new (...args: any[]) => T): string[] {
  return Object.getOwnPropertyNames(type.prototype).filter(
    (propName) => propName !== "constructor" && typeof type.prototype[propName] === "function",
  );
}

const expectation = {
  not: {},
  inChannel: {},
  inGuild: {},
};

function createTestFunction(classType: any, params: MacherContructorArgs, functionName: string) {
  return (...args: any[]) => (new classType(params) as any)[functionName](...args);
}

function createTestsFromMatches<T>(
  names: string[],
  params: MacherContructorArgs,
  classType: new (...args: any[]) => T,
): [string, (...args: any[]) => any][] {
  return names.map((name) => {
    return [name, createTestFunction(classType, params, name)];
  });
}

function set(from: [string, (...args: any[]) => any][], to: any) {
  from.forEach((test) => {
    to[test[0]] = test[1];
  });
}

const messageTestNames = getMessageMatchers();
const roleMatchers = getRoleMatchers();

const _expect: any = <T extends (() => number | string) | number | string>(commandName: T) => {
  const messageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: false },
    MessageMatches,
  );

  const isNotMessageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: true },
    MessageMatches,
  );

  const roleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: false },
    RoleMatches,
  );

  const isNotRoleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: true },
    RoleMatches,
  );

  set(messageTests, expectation);
  set(isNotMessageTests, expectation.not);

  set(roleTests, expectation);
  set(isNotRoleTests, expectation.not);

  expectation.inChannel = (channelId: string) => {
    const channelMatchers: any = {
      not: {},
    };

    const messageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId },
      MessageMatches,
    );

    const isNotmessageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId, isNot: true },
      MessageMatches,
    );

    set(messageTests, channelMatchers);
    set(isNotmessageTests, channelMatchers.not);

    return channelMatchers;
  };

  expectation.inGuild = (guildId: string) => {
    const guildMatchers: any = {
      not: {},
    };

    const roleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    const isNotRoleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    set(roleTests, guildMatchers);
    set(isNotRoleTests, guildMatchers.not);

    return guildMatchers;
  };

  return expectation;
};

const messageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: false, isCascade: true },
  MessageMatches,
);

const isNotMessageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: true, isCascade: true },
  MessageMatches,
);

const roleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: false, isCascade: true },
  RoleMatches,
);

const isNotRoleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: true, isCascade: true },
  RoleMatches,
);

_expect.not = {};

set(messageTests, _expect);
set(isNotMessageTests, _expect.not);

set(roleTests, _expect);
set(isNotRoleTests, _expect.not);

const expect = _expect as Expect;

export { expect };
