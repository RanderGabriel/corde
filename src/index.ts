/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
