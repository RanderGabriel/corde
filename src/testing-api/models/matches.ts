import { MessageEmbed } from 'discord.js';

/**
 * Defines all functions that can be used
 * to check a bot reaction of a command.
 */
export interface Matches {
  /**
   * Defines the message expected to be returned by a
   * command.
   *
   * @param expect A message returned by a bot after invoke a command
   *
   */
  mustReturn(expect: string | MessageEmbed): void;
  /**
   * Defines reactions that must be add to command message.
   *
   * @param reaction Single or list of reactions that must be added to an message
   *
   * @see For how to react message -> https://discordjs.guide/popular-topics/reactions.html#reacting-to-messages
   *
   * @example
   *
   *  bot.on('message', async (message) => {
   *    if (command === 'emoji') {
   *       msg.react('😄');
   *    } else if(command === 'emojis') {
   *       Promise.all([msg.react('😄'), msg.react('🍊')]);
   *    }
   *  });
   *
   *  Tests:
   *
   *  command('emoji').mustAddReaction('😄');
   *  command('emojis').mustAddReaction('😄', '🍊');
   */
  mustAddReaction(...reaction: string[]): void;
}
