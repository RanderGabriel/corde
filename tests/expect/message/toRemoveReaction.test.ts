import MockDiscord from "../../mocks/mockDiscord";
import { initCordeClientWithChannel } from "../../testHelper";
import { Client } from "discord.js";
import { ToRemoveReaction } from "../../../src/expect/matches";
import { TestReport } from "../../../src/types";
import { buildReportMessage, typeOf } from "../../../src/utils";
import { MockEvents } from "../../mocks/mockEvents";
import { runtime } from "../../../src/common";
import { TimeoutError } from "../../../src/errors";

describe("testing toRemoveReaction function", () => {
  let mockDiscord = new MockDiscord();
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should fail due to invalid messageData (number)", async () => {
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);
    // @ts-ignore
    const report = await toRemoveReaction.action(["1"], 1);

    const expectedReport: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expect: message data to be null, undefined, string or an object with id or text properties\n`,
        `received: ${typeOf(1)}`,
      ),
    };

    expect(report).toEqual(expectedReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to invalid emoji (undefined)", async () => {
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);
    // @ts-ignore
    const report = await toRemoveReaction.action(undefined);

    const expectedReport: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: emojis parameter to be an array with string or objects\n`,
        `received: ${typeOf(undefined)}`,
      ),
    };

    expect(report).toEqual(expectedReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to invalid emoji (null)", async () => {
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);
    // @ts-ignore
    const report = await toRemoveReaction.action(null);

    const expectedReport: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: emojis parameter to be an array with string or objects\n`,
        `received: ${typeOf(null)}`,
      ),
    };

    expect(report).toEqual(expectedReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to invalid emoji (object)", async () => {
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);
    // @ts-ignore
    const report = await toRemoveReaction.action({});

    const expectedReport: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: emojis parameter to be an array with string or objects\n`,
        `received: ${typeOf({})}`,
      ),
    };

    expect(report).toEqual(expectedReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a passed test with isNot = false", async () => {
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageReactionsRemove();

    cordeClient.sendTextMessage = jest.fn().mockReturnValue(mockDiscord.message);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);

    const reportModel: TestReport = {
      pass: true,
    };

    const report = await toRemoveReaction.action([mockDiscord.messageReaction.emoji.name]);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a passed test with isNot = true", async () => {
    runtime.setConfigs({ timeOut: 10 }, true);
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    cordeClient.sendTextMessage = jest.fn().mockReturnValue(mockDiscord.message);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", true);

    const reportModel: TestReport = {
      pass: true,
    };

    const report = await toRemoveReaction.action([mockDiscord.messageReaction.emoji.name]);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test with isNot = false", async () => {
    runtime.setConfigs({ timeOut: 10 }, true);
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    cordeClient.sendTextMessage = jest.fn().mockReturnValue(mockDiscord.message);
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: to remove reactions ${[mockDiscord.messageReaction.emoji.name].join(", ")}\n`,
        `received: no reaction was removed to message`,
      ),
    };

    const report = await toRemoveReaction.action([mockDiscord.messageReaction.emoji.name]);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test with isNot = false and timeout", async () => {
    runtime.setConfigs({ timeOut: 10 });
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    cordeClient.sendTextMessage = jest.fn().mockReturnValue(mockDiscord.message);
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageReactionsRemoveToReject();
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: to remove reactions ${[mockDiscord.messageReaction.emoji.name].join(", ")}\n`,
        `received: ${[mockDiscord.messageReaction.emoji.name].join(", ")}`,
      ),
    };

    const report = await toRemoveReaction.action([mockDiscord.messageReaction.emoji.name]);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a passed test with isNot = false", async () => {
    runtime.setConfigs({ timeOut: 10 });
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    cordeClient.sendTextMessage = jest.fn().mockReturnValue(mockDiscord.message);
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageReactionsRemove();
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", false);

    const reportModel: TestReport = {
      pass: true,
    };

    const report = await toRemoveReaction.action([mockDiscord.messageReaction.emoji.name]);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test with isNot = true", async () => {
    runtime.setConfigs({ timeOut: 10 });
    const cordeClient = initCordeClientWithChannel(mockDiscord, new Client(), 1000);
    cordeClient.sendTextMessage = jest.fn().mockReturnValue(mockDiscord.message);
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageReactionsRemove();
    const toRemoveReaction = new ToRemoveReaction(cordeClient, "remove", true);

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: not to remove reactions ${[mockDiscord.messageReaction.emoji.name].join(
          ", ",
        )}\n`,
        `received: ${[mockDiscord.messageReaction.emoji.name].join(", ")}`,
      ),
    };

    const report = await toRemoveReaction.action([mockDiscord.messageReaction.emoji.name]);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });
});
