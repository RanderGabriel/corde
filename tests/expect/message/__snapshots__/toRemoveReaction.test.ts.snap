// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toRemoveReaction function should fail due to invalid emoji (null) 1`] = `
Object {
  "message": "    
    expected: emojis parameter to be an array with string or objects
    received: null
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should fail due to invalid emoji (object) 1`] = `
Object {
  "message": "    
    expected: emojis parameter to be an array with string or objects
    received: object
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should fail due to invalid emoji (undefined) 1`] = `
Object {
  "message": "    
    expected: emojis parameter to be an array with string or objects
    received: undefined
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should fail due to invalid messageIdentifier (number) 1`] = `
Object {
  "message": "    
    expect: message data to be null, undefined, string or an object with id or text properties
    received: number
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should return a failed test with isNot = false 1`] = `
Object {
  "message": "    
    expected: to remove reactions ðŸ˜€
    received: no reaction was removed to message
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should return a failed test with isNot = false and timeout 1`] = `
Object {
  "message": "    
    expected: to remove reactions ðŸ˜€
    received: ðŸ˜€
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should return a failed test with isNot = true 1`] = `
Object {
  "message": "    
    expected: not to remove reactions ðŸ˜€
    received: ðŸ˜€
",
  "pass": false,
}
`;

exports[`testing toRemoveReaction function should return a passed test with isNot = false 1`] = `
Object {
  "pass": true,
}
`;

exports[`testing toRemoveReaction function should return a passed test with isNot = false 2`] = `
Object {
  "pass": true,
}
`;

exports[`testing toRemoveReaction function should return a passed test with isNot = true 1`] = `
Object {
  "pass": true,
}
`;
